android/tangram/src/com/mapzen/tangram/MapData.java

```diff
@@ -1,161 +1,154 @@
+import com.mapzen.tangram.geometry.Polyline;
+
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Map;
 
-public class MapData extends DataSource {
-    private transient long swigCPtr;
-    private boolean swigCMemOwnDerived;
+/**
+ * {@code MapData} is a named collection of drawable map features.
+ */
+public class MapData {
 
-    protected MapData(long cPtr, boolean cMemoryOwn) {
-        super(TangramJNI.MapData_SWIGSmartPtrUpcast(cPtr), true);
-        swigCMemOwnDerived = cMemoryOwn;
-        swigCPtr = cPtr;
-    }
+    String name;
+    long pointer = 0;
+    MapController map;
+    List<String> geojson = new ArrayList<>();
+    List<Geometry> features = new ArrayList<>();
 
-    protected static long getCPtr(MapData obj) {
-        return (obj == null) ? 0 : obj.swigCPtr;
+    /**
+     * Construct a collection of drawable map features.
+     * @param name The name of the data collection. Once added to a map, features from this
+     * {@code MapData} will be available from a data source with this name, just like a data source
+     * specified in a scene file.
+     */
+    public MapData(String name) {
+        this.name = name;
     }
 
-    protected void finalize() {
-        delete();
+    /**
+     * Get the name of this {@code MapData}.
+     * @return The name.
+     */
+    public String name() {
+        return name;
     }
 
-    public synchronized void delete() {
-        if (swigCPtr != 0) {
-            if (swigCMemOwnDerived) {
-                swigCMemOwnDerived = false;
-                TangramJNI.delete_MapData(swigCPtr);
-            }
-            swigCPtr = 0;
+    /**
+     * Add the features from this {@code MapData} to a map.
+     * <p>
+     * This {@code MapData} will be associated with the given map until {@link #removeFromMap()}
+     * is called or until this method is called again.
+     * @param map The {@code MapController} managing the destination map.
+     */
+    public void addToMap(MapController map) {
+        if (map == null) {
+            throw new RuntimeException("Cannot assign a null map, use removeFromMap() instead");
+        }
+        if (this.map != null) {
+            removeFromMap();
         }
-        super.delete();
+        this.map = map;
+        this.pointer = map.nativeAddDataSource(name);
+        syncWithMap();
     }
 
-
     /**
-     * Construct a new MapData object for adding drawable data to the map
-     * @param name Name of the data source in the scene file for styling this
-     *             object's data
+     * Remove this {@code MapData} from the map it is currently associated with. This must not be
+     * called when this {@code MapData} is not associated with a map.
      */
-    public MapData(String name) {
-        this(TangramJNI.new_MapData(name, ""), true);
+    public void removeFromMap() {
+        if (map == null) {
+            throw new RuntimeException("There is no associated map for this MapData");
```

Comment:
Calling `removeFromMap()` when there is no map associated is a programmer error. The idea behind throwing an exception here is for the code to teach and enforce correct management of `MapData` instead of permitting confusing sloppiness.